File: programmers/js/new/level2/가장_큰_정사각형_찾기.js
function solution(board) {
    const arr = Array.from(Array(board.length), () => Array(board[0].length).fill(0));
    for (let j = 0; j < board[0].length; ++j) {
        arr[0][j] = board[0][j];
    }
    for (let i = 0; i < board.length; ++i) {
        arr[i][0] = board[i][0];
    }
    let max = arr[0][0] === 0 ? 0 : 1;
    for (let i = 1; i < board.length; ++i) {
        for (let j = 1; j < board[0].length; ++j) {
            if (board[i][j] === 1) {
                arr[i][j] = Math.min(arr[i][j - 1], arr[i - 1][j], arr[i - 1][j - 1]) + 1;
                if (arr[i][j] > max) {
                    max = arr[i][j];
                }
            }
        }
    }

    return max * max;
}

console.log(
    solution([
        [0, 1, 1, 1],
        [1, 1, 1, 1],
        [1, 1, 1, 1],
        [0, 0, 1, 0],
    ])
);

console.log(
    solution([
        [0, 0, 1, 1],
        [1, 1, 1, 1],
    ])
);

File: programmers/python/level2/가장_큰_정사각형_찾기.py
def solution(board):
    answer = 0
    # 왼쪽, 위쪽, 대각선 위쪽 에서부터 연속된 1 개수
    count_list = [[0 for c in range(len(board[r]))] for r in range(len(board))]
    count_list[0][0] = 1 if board[0][0] == 1 else 0
    for c in range(1,len(board[0])):
        count_list[0][c] = board[0][c]
    for r in range(1,len(board)):
        count_list[r][0] = board[r][0]
    answer = 1 if count_list[0][0] else 0
    for r in range(1, len(board)):
        for c in range(1, len(board[r])):
            if board[r][c] == 1:
                count_list[r][c] = min(count_list[r][c-1], count_list[r-1][c], count_list[r-1][c-1]) + 1
                answer = max(answer, count_list[r][c] ** 2)
    return answer
